### ISSUE TITLE
How to use weighted FTS with multiple fields of similar weight

### ISSUE THREAD

**Issue by NeelDigonto** [mikro-orm/mikro-orm#5174]
How to use weighted FTS with multiple fields of similar weight?

For example how to achieve something like this?

```sql
BEGIN;

CREATE TABLE log(
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  level TEXT NOT NULL,
  message TEXT NOT NULL,
  traceid TEXT NOT NULL,
  document TSVECTOR GENERATED ALWAYS AS (
    setweight(to_tsvector('english', level), 'A') ||
    setweight(to_tsvector('english', traceid), 'A') ||
    setweight(to_tsvector('english', message), 'B')
  ) STORED
);

CREATE INDEX log_document_idx ON log USING GIN (document);

COMMIT;
```

This query has set both `level` and `traceid` to weight `A`.
Reference: 
https://www.postgresql.org/docs/current/textsearch-controls.html

I have gone over the MikroORM documentation.

```ts
type WeightedFullTextValue = {
    A?: string | null | undefined;
    B?: string | null | undefined;
    C?: string | null | undefined;
    D?: string | null | undefined;
}

export type WeightedFullTextValue = {
    [K in FullTextWeight]?: string | null;
};
```

Here I can only set one column for each weighted level.

If the requested feature is not possible in the short term, I am also fine with `stored procedures`, and I will also prefer that to the `onUpdate` way of doing this, since I could see seperate `UPDATE` queries being sent to db with the `onUpdate` method.

I tried the following with STORED procedures.

```ts
@Entity()
export class Tenant {
  @PrimaryKey({ type: 'integer' })
  id!: number;

  @Property({ type: 'text', nullable: false })
  name: string = '';

  @Property({ type: 'text', nullable: false })
  description = '';

  @Property({ type: 'integer', nullable: false })
  maxActiveUsers = 20;

  @Property({ type: 'timestamptz' })
  createdAt = new Date();

  @Property({ type: 'timestamptz', onUpdate: () => new Date() })
  updatedAt = new Date();
  @Index({ type: "GIN" })
  @Property({
    columnType: `TSVECTOR GENERATED ALWAYS AS setweight(to_tsvector('english', level), 'A') ||
    setweight(to_tsvector('english', traceid), 'A') ||
    setweight(to_tsvector('english', message), 'B') STORED NOT`,
    type: 'unknown',
    nullable: true
  })
  searchableTitle!: string;
}
```
The last `NOT` is there since drizzle puts a `null` attribute due to `nullable: true`.
I kept the `nullable: true` since I it's a stored procedure and I don't intend to populate that field, but only query it.

But it fails with 

```sh
[query] begin
[query] insert into "Tenant" ("name", "description", "maxActiveUsers", "createdAt", "updatedAt") values ('yayma', 'tomdo', 20, '2024-01-26T16:21:11.933Z', '2024-01-26T16:21:11.933Z') returning "id" [took 2 ms, 1 row affected]
[query] update "Tenant" set "searchableTitle" = to_tsvector('simple', '''tomdo'':2B ''yayma'':1A'), "updatedAt" = '2024-01-26T16:21:11.937Z' where "id" = 1 returning "searchableTitle" [took 2 ms]
[query] rollback
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

DriverException: update "Tenant" set "searchableTitle" = to_tsvector('simple', '''tomdo'':2B ''yayma'':1A'), "updatedAt" = '2024-01-26T16:21:11.937Z' where "id" = 1 returning "searchableTitle" - column "searchableTitle" can only be updated to DEFAULT
```

I tried the generated field too, but something similar happened there too.

Can you give some alternative for getting it to work?

**Comment by B4nan** [comment]
any idea on this @jsprw?

**Comment by NeelDigonto** [comment]
@B4nan 
Also, with something simplier and documented.

Lets say 
```ts
@Entity()
export class Tenant {
  @PrimaryKey({ type: 'integer' })
  id!: number;

  @Property({ type: 'text' })
  name: string = '';

  @Property({ type: 'text' })
  description = '';

  @Property({ type: 'integer' })
  maxActiveUsers = 20;

  @Property({ type: 'timestamptz' })
  createdAt = new Date();

  @Property({ type: 'timestamptz', onUpdate: () => new Date() })
  updatedAt = new Date();

  @Index({ type: 'fulltext' })
  @Property({
    type: FullTextType,
    onCreate: (tenant: Tenant) => ({ A: tenant.name, B: tenant.description }),
    onUpdate: (tenant: Tenant) => ({ A: tenant.name, B: tenant.description })
  })
  searchableTitle!: WeightedFullTextValue;
```

When I do something like 
```ts
  const rnum = Math.floor(Math.random() * 1_000_000)

  const tenant = new Tenant();
  tenant.name = `org-${rnum}`;
  tenant.description = `comment-${rnum}`;

  await lem.persist(tenant).flush();
```

We get 
```sh
[query] begin
[query] insert into "Tenant" ("name", "description", "maxActiveUsers", "createdAt", "updatedAt", "searchableTitle") values ('org-274573', 'comment-274573', 20, '2024-01-26T20:13:05.418Z', '2024-01-26T20:13:05.418Z', setweight(to_tsvector('simple', 'org-274573'), 'A') || setweight(to_tsvector('simple', 'comment-274573'), 'B')) returning "id" [took 9 ms, 1 row affected]
[query] commit
```

But by adding another find query, the update statement is kind of messed up?

```ts
  const [items, total] = await lem.findAndCount(
    Tenant,
    {},
    {
      limit: 10,
      offset: 0,
    },
  );
  console.log(items, total);

  const rnum = Math.floor(Math.random() * 1_000_000)

  const tenant = new Tenant();
  tenant.name = `org-${rnum}`;
  tenant.description = `comment-${rnum}`;

  await lem.persist(tenant).flush();
```

```sh
[query] begin
[query] insert into "Tenant" ("name", "description", "maxActiveUsers", "createdAt", "updatedAt", "searchableTitle") values ('org-694169', 'comment-694169', 20, '2024-01-26T20:17:31.031Z', '2024-01-26T20:17:31.031Z', setweight(to_tsvector('simple', 'org-694169'), 'A') || setweight(to_tsvector('simple', 'comment-694169'), 'B')) returning "id" [took 6 ms, 1 row affected]
[query] update "Tenant" set "searchableTitle" = case when ("id" = 7) then setweight(to_tsvector('simple', 'org-364372'), 'A') || setweight(to_tsvector('simple', 'comment-364372'), 'B') when ("id" = 1) then setweight(to_tsvector('simple', 'org-424799'), 'A') || setweight(to_tsvector('simple', 'comment-424799'), 'B') when ("id" = 2) then setweight(to_tsvector('simple', 'org-863365'), 'A') || setweight(to_tsvector('simple', 'comment-863365'), 'B') when ("id" = 3) then setweight(to_tsvector('simple', 'org-267250'), 'A') || setweight(to_tsvector('simple', 'comment-267250'), 'B') when ("id" = 4) then setweight(to_tsvector('simple', 'org-20977'), 'A') || setweight(to_tsvector('simple', 'comment-20977'), 'B') when ("id" = 5) then setweight(to_tsvector('simple', 'org-769056'), 'A') || setweight(to_tsvector('simple', 'comment-769056'), 'B') when ("id" = 6) then setweight(to_tsvector('simple', 'org-289869'), 'A') || setweight(to_tsvector('simple', 'comment-289869'), 'B') when ("id" = 8) then setweight(to_tsvector('simple', 'org-274573'), 'A') || setweight(to_tsvector('simple', 'comment-274573'), 'B') else "searchableTitle" end, "updatedAt" = case when ("id" = 7) then '2024-01-26T20:17:31.036Z' when ("id" = 1) then '2024-01-26T20:17:31.036Z' when ("id" = 2) then '2024-01-26T20:17:31.036Z' when ("id" = 3) then '2024-01-26T20:17:31.036Z' when ("id" = 4) then '2024-01-26T20:17:31.036Z' when ("id" = 5) then '2024-01-26T20:17:31.036Z' when ("id" = 6) then '2024-01-26T20:17:31.037Z' when ("id" = 8) then '2024-01-26T20:17:31.037Z' else "updatedAt" end where "id" in (7, 1, 2, 3, 4, 5, 6, 8) returning "searchableTitle" [took 6 ms, 8 rows affected]
[query] commit
```
![image](https://github.com/mikro-orm/mikro-orm/assets/57017288/a8b8dadf-2eda-4c0b-b935-c70e4dff35c3)

Why is this updating the `searchableTitle` field of all the entities, except the new entity everything is unchanged.
For these kind of complex data types `stored procedures` will be much better ig.
So, if possible I just want `mikro-orm` to know that I will provide a `generated` | `stored procedure` to the db and the db will populate the field on its own.

So mikro-orm should only touch this field while migrating and doing search querries.
Is this something which can do or is it better to let `mikro-orm` handle it.

But anyways this approach is kindof unstable if many rows have same data, then the db `searchableTitle` field becomes a mess.
For example if I just remove the `onUpdate` but keep the `onCreate` field.

![image](https://github.com/mikro-orm/mikro-orm/assets/57017288/76b92e29-f01c-40b0-9bed-b2268b1ab653)

**Comment by jsprw** [comment]
In your example, I see you're using properties that are not defined in your entity.

However, when you add these properties to the entity, you can combine them into a string for weighted value A. You could combine them with spaces in between. If I'm not mistaken the two statements below yield the same result.

```sql
select setweight(to_tsvector('english', level), 'A') || setweight(to_tsvector('english', traceid), 'A');
select setweight(to_tsvector('english', level || ' ' || traceid), 'A');

---standalone test
select setweight(to_tsvector('english', 'HELLO'), 'A') || setweight(to_tsvector('english', 'NAME'), 'A');
select setweight(to_tsvector('english', 'HELLO' || ' ' || 'NAME'), 'A');
```


```typescript
@Entity()
export class Tenant {
  @PrimaryKey({ type: 'integer' })
  id!: number;

  @Property({ type: 'text', nullable: false })
  name: string = '';

  @Property({ type: 'text', nullable: false })
  level: string = '';

  @Property({ type: 'text', nullable: false })
  traceid: string = '';

  @Property({ type: 'text', nullable: false })
  description = '';

  @Property({ type: 'integer', nullable: false })
  maxActiveUsers = 20;

  @Property({ type: 'timestamptz' })
  createdAt = new Date();

  @Property({ type: 'timestamptz', onUpdate: () => new Date() })
  updatedAt = new Date();

  @Index({ type: "fulltext" })
  @Property({
    type: new FullTextType('english'),
    nullable: true,
    onUpdate: (tenant: Tenant) => ({ A: [tenant.name, tenant.traceid, tenant. level].join(' '), B: tenant.message })
  })
  searchableTitle!: string;
}
```

Can you check whether the above example works for you?

> For these kind of complex data types stored procedures will be much better ig.
Yes, but in this example you can control the the input to the ts_vector from the JS backend (eg add additional search terms), while with stored procedures, the logic is quite hidden and you can only access database fields.

> So mikro-orm should only touch this field while migrating and doing search querries.
We have been experiencing this as well. It occurs due to the ORM not being able to compare the values (the input value is a lot different than the output value) and therefore thinking the value is always dirty when you've changed it once. When it is removed and re-added to the uow, it works fine. As you can see in the insert query, nothing special happens there

Why the field looks so messed up, is something that is not caused by the queries above your screenshot. If you execute those functions in a single select, you will get the expected result of the to_tsvector function.

**Comment by NeelDigonto** [comment]
@jsprw I tried it 

```ts
@Entity()
export class Tenant {
  @PrimaryKey({ type: 'integer' })
  id!: number;

  @Property({ type: 'text' })
  name: string = '';

  @Property({ type: 'text' })
  kind = '';

  @Property({ type: 'text' })
  description = '';


  @Property({ type: 'integer' })
  maxActiveUsers = 20;

  @Property({ type: 'timestamptz' })
  createdAt = new Date();

  @Property({ type: 'timestamptz', onUpdate: () => new Date() })
  updatedAt = new Date();

  @Index({ type: "fulltext" })
  @Property({
    type: new FullTextType('english'),
    nullable: true,
    onUpdate: (tenant: Tenant) => ({ A: [tenant.name, tenant.kind].join(' '), B: tenant.description })
  })
  searchableTitle!: string;
}
```

```ts
  const rnum = Math.floor(1_000_000)

  const tenant = new Tenant();
  tenant.name = `org`;
  tenant.kind = `gold`;
  tenant.description = `comment`;

  await lem.persist(tenant).flush();
```
[query] begin
[query] insert into "Tenant" ("name", "kind", "description", "maxActiveUsers", "createdAt", "updatedAt") values ('org', 'gold', 'comment', 20, '2024-01-26T21:49:29.328Z', '2024-01-26T21:49:29.328Z') returning "id" [took 2 ms, 1 row affected]
[query] commit
```sh

```

Without onCreate, the `tsvector` field remains null.

![image](https://github.com/mikro-orm/mikro-orm/assets/57017288/b07d88ae-c72a-4421-a8ec-1a941d74eebe)

**Comment by jsprw** [comment]
> As soon as I add an onCreate with the same value as on onUpdate and the field gets populated,
it sends a complex update querry.

Is that query correct?

**Comment by NeelDigonto** [comment]
The query is correct but quite massive as I showed an example earlier.

The DB state seems okay though

![image](https://github.com/mikro-orm/mikro-orm/assets/57017288/88054bfc-8596-46d8-b976-5d69d59a79ed)

**Comment by jsprw** [comment]
> With your example there was no update query going on, after adding in onCreate the query is somewhat like

Yeah sorry, the onCreate was meant to be there but you've added it correctly. It looks like the original issue was resolved by this change?

>  The db state becomes okay, but just that the update query is a bit massive.

Yeah that has to do with what I described before. When you don't load all these entities in the UoW, it wont generate this big of a query (although the query doesn't do any harm and updates the right fields correctly).
